### PR TITLE
Issue #12645: All GitHub CI workflows should have concurrency of 1

### LINKED ISSUE #12645: All Github CI workflows should have concurrency of 1
Currently, all of our Github workflows will continue to run after new pushes to a PR. This is usually pointless, and consumes a lot of extra resources (credits + occupied containers). This causes other workflows to be queued for long periods, and also can slow down the release process.

Solution: we need to set all github workflows to have a concurrency of 1, and automatically cancel in progress workflows.

Example of yaml update:
```
concurrency:
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
  cancel-in-progress: true
```

We also need to create some check to make sure that all github workflows have this concurrency setting.

### PR DESCRIPTION
**Pull request by nrmancuso**
Closes #12645

Example of failure: https://circleci.com/gh/checkstyle/checkstyle/234145

### CODE REVIEW

File: .github/workflows/update_sources.yml [hunk 1]
```diff
@@ -9,6 +9,11 @@ on:
       version:
         description: 'Target Version without (-SNAPSHOT)'
         required: true
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
```
Comment (nrmancuso) [inline .github/workflows/update_sources.yml:15]: Github CI uses concurrency groups to determine concurrency. So, we need to create a group name that is unique to each PR and each workflow in that PR. However, when it comes to tags and master, `github.event.pull_request.number` will not exist. `||` allow us to specify an alternative, which in this case is the fully formed `ref`.

**Review by romani — APPROVED** [review]
ok to merge

**Comment by rnveach** [comment]
@nrmancuso CI is failing

File: .ci/validation.sh [hunk 2]
```diff
@@ -1105,6 +1105,27 @@ ci-temp-check)
     exit $fail
   ;;
 
+  check-github-workflows-concurrency)
+    GITHUB_WORKFLOW_FILES=$(find .github/workflows -maxdepth 1 -not -type d -name "*.y*ml")
```
Comment (rnveach) [inline .ci/validation.sh:1109]: > "*.y*ml"

Is this to include `yaml` files too? I don't see any changes in this PR for `yaml`.

**Comment by nrmancuso** [comment]
> @nrmancuso CI is failing

Yes, this is intentional :)

File: .ci/validation.sh (hunk 2 above)
Comment (romani) [reply .ci/validation.sh:1109]: I do not see any "yaml" at https://github.com/checkstyle/checkstyle/tree/master/.github/workflows

Comment (nrmancuso) [reply .ci/validation.sh:1109]: Yes, just future-proofing in case someone uses yaml extension.

**Review by rnveach — APPROVED** [review]

**Pull request merged**